infor-design/enterprise-wc
# 135 - IdsPopup Layout Logic and Improvements

**Explain the details for making this change. What existing problem does the pull request solve?**
This PR makes some general improvements to the IdsPopup component:
- Fixes a bug reported where the arrow applied to some popups would visually "jump" to the correct position after the Popup was rendered.
- Refactors the lifecycle overall to not call `refresh()` as often as before, improving speed and cutting down on DOM repaints.
- Removes `safeSetAttribute/safeRemoveAttribute` in favor of better checks on setter values to prevent double-setting of properties from occurring.
- Adds a `containingElem` setting that can be configured to keep a Popup confined within the boundaries of another element on the page.  By default, this is the IdsContainer present on all pages, so all Popups will remain within those boundaries.
- Adds a `bleed` setting, which can be used to disable/enable the containment behavior.
- Ports over the `nudge` and `flip` behaviors from the 4.x Place component, and directly integrates them into the `placeAtCoords` and `placeAgainstParent` routines.

**Related github/jira issue (required)**:
Closes #135 
Closes #198 

**Steps necessary to review your pull request (required)**:

_To test #198:_
- Open http://localhost:4300/ids-menu-button 
- Open both menus and watch the arrow.  It should open and remain in the correct placement, beneath each component.

_To test containment:_
- Open http://localhost:4300/ids-popup-menu
- Right-click (or CTRL+Click) with the mouse cursor placed closely to the edge of the page  An IdsPopupMenu should open, and should not be cut off by the browser viewport.
- Open the submenus, and the same should be true for each.

_Another containment test:_
- Open http://localhost:4300/ids-popup/test-sandbox
- Under the "Alignment Targets" fieldset, click "None (Use Coordinates)".  The Popup will move to the top/left corner of the browser viewport.
- Under the "Containment" fieldset, click "Test Container".  The Popup will move to the top/left of the gray test container element in the page.
- Under the "Coordinates" fieldset, set the X/Y coordinates each to at least 600px.  The Popup will move to the bottom/right corner of the gray test container, but it will not cross the edges.
- Under the "Other" fieldset, click "Allow Bleeds".  The Popup will reposition and cross the edges of the gray test container.

**Included in this Pull Request**:
